src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,739 @@
+    else
+    {
+        // Re-evaluate our routing policy and send Router Advertisement if necessary.
+        EvaluateRoutingPolicy();
+    }
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<RoutingManager>().HandleDiscoveredOnLinkPrefixInvalidTimer();
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(void)
+{
+    mDiscoveredOnLinkPrefix.Clear();
+
+    // The discovered on-link prefix becomes invalid, start Router Solicitation
+    // to discover new one.
+    StartRouterSolicitation();
+}
+
+void RoutingManager::HandleRouterSolicit(const Ip6::Address &aSrcAddress,
+                                         const uint8_t *     aBuffer,
+                                         uint16_t            aBufferLength)
+{
+    OT_UNUSED_VARIABLE(aSrcAddress);
+    OT_UNUSED_VARIABLE(aBuffer);
+    OT_UNUSED_VARIABLE(aBufferLength);
+
+    otLogInfoBr("received Router Solicitation from %s on interface %s", aSrcAddress.ToString().AsCString(),
+                GetInfraIfName());
+
+    mRouterAdvertisementTimer.Start(Random::NonCrypto::GetUint32InRange(0, kMaxRaDelayTime));
+}
+
+bool RoutingManager::HandlePrefixInfoOption(const RouterAdv::PrefixInfoOption &aPio)
+{
+    bool      needReevaluate = false;
+    TimeMilli expireTime;
+
+    if (aPio.GetValidLifetime() == 0)
+    {
+        if (mDiscoveredOnLinkPrefix.IsEqual(aPio.GetPrefix(), aPio.GetPrefixLength()))
+        {
+            otLogInfoBr("invalidate discovered on-link prefix: %s", mDiscoveredOnLinkPrefix.ToString().AsCString());
+            mDiscoveredOnLinkPrefix.Clear();
+            mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+            StartRouterSolicitation();
+        }
+        ExitNow();
+    }
+
+    expireTime = GetPrefixExpireTime(aPio.GetValidLifetime());
+
+    if (!IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix) ||
+        (mDiscoveredOnLinkPrefixInvalidTimer.IsRunning() &&
+         expireTime > mDiscoveredOnLinkPrefixInvalidTimer.GetFireTime()))
+    {
+        mDiscoveredOnLinkPrefix.Set(aPio.GetPrefix(), aPio.GetPrefixLength());
+
+        otLogInfoBr("set discovered on-link prefix to %s, valid lifetime: %u seconds",
+                    mDiscoveredOnLinkPrefix.ToString().AsCString(), aPio.GetValidLifetime());
+
+        if (aPio.GetValidLifetime() == RouterAdv::kInfiniteLifetime)
+        {
+            mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+        }
+        else
+        {
+            mDiscoveredOnLinkPrefixInvalidTimer.FireAt(expireTime);
+        }
+
+        mRouterSolicitTimer.Stop();
+        needReevaluate = true;
+    }
+
+exit:
+    return needReevaluate;
+}
+
+TimeMilli RoutingManager::GetPrefixExpireTime(uint32_t aValidLifetime)
+{
+    if (aValidLifetime * 1000ULL + TimerMilli::GetNow().GetValue() < TimeMilli::kMaxDuration)
+    {
+        return Time(aValidLifetime * 1000 + TimerMilli::GetNow().GetValue());
+    }
+    else
+    {
+        return Time(Time::kMaxDuration);
```

Comment:
Is it meaningful to convert a uint32 value to `Time` directly?

Why not just return a uint32_t duration for this method?